server/src/main/java/io/crate/analyze/ArraySliceValidator.java

```diff
@@ -0,0 +1,171 @@
+            context.setBase(node);
+            return null;
+        }
+
+        @Override
+        protected Void visitExpression(Expression node, ArraySliceContext context) {
+            throw new UnsupportedOperationException(
+                String.format(Locale.ENGLISH,
+                              "An expression of type %s cannot have an array slice ([<from>:<to>])",
+                              node.getClass().getSimpleName())
+            );
+        }
+    }
+
+    private static class ArraySliceIndexVisitor extends AstVisitor<Void, ArraySliceContext> {
+
+        private static final ArraySliceIndexVisitor FROM_INSTANCE = new ArraySliceIndexVisitor(
+            ArraySliceContext::setFrom,
+            ArraySliceContext::getFrom
+        );
+        private static final ArraySliceIndexVisitor TO_INSTANCE = new ArraySliceIndexVisitor(
+            ArraySliceContext::setTo,
+            ArraySliceContext::getTo
+        );
+
+        private final BiConsumer<ArraySliceContext, Optional<Expression>> setter;
+        private final Function<ArraySliceContext, Optional<Expression>> getter;
+
+        public ArraySliceIndexVisitor(BiConsumer<ArraySliceContext, Optional<Expression>> setter,
+                                      Function<ArraySliceContext, Optional<Expression>> getter) {
+            this.setter = setter;
+            this.getter = getter;
+        }
+
+        private static void raiseInvalidIndexValue() {
+            throw new UnsupportedOperationException(
+                String.format(Locale.ENGLISH, "Array index must be in range 1 to %s", MAX_VALUE)
+            );
+        }
+
+        private void validateNestedArraySlice(ArraySliceContext context) {
+            if (getter.apply(context).isPresent()) {
+                throw new UnsupportedOperationException("Nested array slice is not supported");
+            }
+        }
+
+        @Override
+        public Void visitParameterExpression(ParameterExpression node, ArraySliceContext context) {
+            throw new UnsupportedOperationException("Parameter substitution is not supported in an array slice");
+        }
+
+        @Override
+        protected Void visitLongLiteral(LongLiteral node, ArraySliceContext context) {
+            validateNestedArraySlice(context);
+            long value = node.getValue();
+
+            if (value < 1 || value > MAX_VALUE) {
+                raiseInvalidIndexValue();
+            }
+            setter.accept(context, Optional.of(new Cast(node, new ColumnType<>("integer"))));
+            return null;
+        }
+
+        @Override
+        protected Void visitIntegerLiteral(IntegerLiteral node, ArraySliceContext context) {
+            validateNestedArraySlice(context);
+            if (node.getValue() < 1) {
+                raiseInvalidIndexValue();
+            }
+            setter.accept(context, Optional.of(new Cast(node, new ColumnType<>("integer"))));
```

Comment:
Is the cast needed here?

Reply:
it works with an integer literal https://github.com/crate/crate/pull/11949/commits/74d2ed1e691196f264d8233e355ae6b772a0a3bf#diff-71c51d29718721f26b416dbb43ec8f67ab62b7a01470f20fc687942fc8ca30eeR143